### ISSUE TITLE
Immutable classes with immutable fields should use immutable implementation

### ISSUE THREAD

**Issue by mjburghoffer** [immutables/immutables#214]
Sometimes when defining schema, there is a level of nested of fields, such that interface A has a non-primitive field that is of type B, and another of type C.  If the implementation of B is mutable, then the immutable version of A is therefore also, unfortunately, mutable.  Another side-effect is `a.getB()` does not have the niceties like `copy()`, etc, provided by Immutables.  This, of course, should be an optional Style defaulted to "off".

Assuming you have a couple classes defined like:
```java
@Value.Immutable
public interface Point {
  int getX();
  int getY();
}

@Value.Immutable
public interface Line {
  Point start();
  Point end();
}
```

The generated `Point` is fine:
```java
public class ImmutablePoint {
  private final int x;
  private final int y;

  public int getX() {
    return x;
  }

  public int getY() {
    return y;
  }
}
```

It would be nice if the generated implementation of `Line` was like:

```java
public class ImmutableLine {
  private final ImmutablePoint start;
  private final ImmutablePoint end;

  public ImmutablePoint getStart() {
    return start;
  }

  public ImmutablePoint getEnd() {
    return end;
  }
}
```

Rather than:

```java
public class ImmutableLine {
  private final Point start;
  private final Point end;

  public Point getStart() {
    return start;
  }

  public Point getEnd() {
    return end;
  }
}
```

**Comment by elucash** [comment]
We thought about it some time ago. In principle this may be of some value. Immediate difficulty arises with generics, i.e. when you use `Optional` or `List`, because `Optional<ImmutablePoint>`  is not compatible with `Optional<Point>`, just as an example. I think we might implement this, but with limitations.
However, what would I suggest is to also explore reverse generation style. It is when your generated classes considered as primary classes to be used in code.

```java
@Target({ElementType.PACKAGE, ElementType.TYPE})
@Retention(RetentionPolicy.CLASS) // Make it class retention for incremental compilation
@Value.Style(
    typeAbstract = "Abstract*", // 'Abstract' prefix will be detected and trimmed
    typeImmutable = "*" // No prefix or suffix for generated immutable type)
public @interface MyStyle {}
...

@Value.Immutable // if no attributes are specified, then defaults will be used
@MyStyle // This annotation could be also placed on package
public interface AbstractPoint {
  @Value.Parameter int getX();
  @Value.Parameter int getY();
}

@MyStyle
@Value.Immutable
public interface AbstractLine {
  Point start();  // Here you specify generated type, and not AbstractPoint
  Point end();
}

// Below, only immutable types are used, not abstract ones
Line line =  Line.builder()
    .start(Point.of(1, 2))
    .end(Point.of(3, 4))
    .build();
Point x = line.getX();
Point y = line.getY();
```

What do you think?

**Comment by mjburghoffer** [comment]
@elucash Thanks for taking the time to respond, I hadn't thought to use the symbol of the yet-to-be-generated class.  Great idea, and I think that covers all the use cases I can think of.

**Comment by elucash** [comment]
Covariant override should work if `Optional<? extends Point>`  is being overridden by `Optional<ImmutablePoint>` etc, but then we are not so good currently at generics/wildcards analysis (and wildcards in return types are recipe for the trouble), so we might omit those cases anyway, returning `ImmutablePoint` only from regular getters.

**Comment by elucash** [comment]
I managed to put together really minimal, "straw-man" implementation for this. We could not afford currently to handle all complicated cases with collections, wildcards and so on (they are not simple, just believe me) so just simple attributes are handled.
I still hope that straightforward approach we used would be of some value.
To get the feeling of what's there, you can take a look at this [sample ...Canvas.java](https://github.com/immutables/immutables/blob/master/value-fixture/src/org/immutables/fixture/deep/Canvas.java). The feature is being enabled by using `deepImmutablesDetection = true` style switch.

Limitations: No dedicated support for using immutable implementation with default, derived, optional and collection attributes.

This is just released in 2.1.1. Thank you for suggestion!
